### PR TITLE
[wip] [Bugfix] ProxyRule pattern in the response should not strip slashes

### PR DESCRIPTION
**Pull request by Martouta**
Closes [THREESCALE-3872](https://issues.jboss.org/browse/THREESCALE-3872)

1. [Refactor] ProxyRuleDecorator#pattern without using 'File'.
2. [Bugfix] ProxyRuleDecorator#pattern should not strip slashes.

### CODE REVIEW

File: app/decorators/proxy_rule_decorator.rb [hunk 1]
```diff
@@ -4,11 +4,9 @@ class ProxyRuleDecorator < ApplicationDecorator
   self.include_root_in_json = false
 
   def pattern
-    pattern_value = object.pattern
-    return pattern_value unless backend_api_path
-    parts = ['/', backend_api_path]
-    parts << pattern_value unless pattern_value == '/'
-    File.join(*parts)
+    path_joined_parts = ['/', backend_api_path, object.pattern].join('/')
```
Comment (guicassolato) [inline app/decorators/proxy_rule_decorator.rb]:
```suggestion
    parts = [backend_api_path, object.pattern].compact
    '/' + parts.map { |part| StringUtils::StripSlash.strip_slash(part).presence }.compact.join('/')
```

File: app/decorators/proxy_rule_decorator.rb [hunk 2]
```diff
@@ -4,11 +4,9 @@ class ProxyRuleDecorator < ApplicationDecorator
...
+    path_without_duplicated_slashes = path_joined_parts.gsub(%r{\/{2,}}, '/')
```
Comment (guicassolato) [inline app/decorators/proxy_rule_decorator.rb]:
```suggestion
```

File: app/decorators/proxy_rule_decorator.rb [hunk 3]
```diff
@@ -4,11 +4,9 @@ class ProxyRuleDecorator < ApplicationDecorator
...
+    path_without_duplicated_slashes.chomp('/').presence || '/'
```
Comment (guicassolato) [inline app/decorators/proxy_rule_decorator.rb]:
```suggestion
```

**Comment by guicassolato** [comment]
```
diff --git a/test/decorators/proxy_rule_decorator_test.rb b/test/decorators/proxy_rule_decorator_test.rb
index cf66c872..989d0a38 100644
--- a/test/decorators/proxy_rule_decorator_test.rb
+++ b/test/decorators/proxy_rule_decorator_test.rb
@@ -21,6 +21,12 @@ class ProxyRuleDecoratorTest < Draper::TestCase
     assert_equal '/hello', decorator.pattern
   end

+  test 'pattern with trailing slash' do
+    proxy_rule = FactoryBot.build_stubbed(:proxy_rule, owner: backend_api, pattern: '/hello/')
+    decorator = proxy_rule.decorate(context: { backend_api_path: 'mybackend' })
+    assert_equal '/mybackend/hello', decorator.pattern
+  end
+
   test 'catch-all pattern for backend with path' do
     proxy_rule = FactoryBot.build_stubbed(:proxy_rule, owner: backend_api, pattern: '/')
     decorator = proxy_rule.decorate(context: { backend_api_path: 'mybackend' })
```

File: app/decorators/proxy_rule_decorator.rb (hunk 1 above)
Comment (duduribeiro) [inline app/decorators/proxy_rule_decorator.rb]: What about starting with `''`. The join will append the first `/`.

Comment (guicassolato) [reply app/decorators/proxy_rule_decorator.rb]: This suggestions is slightly different from what we had before. It removes the trailing slash in all cases, but the one where the mapping rule pattern is nothing but `/`.

Comment (duduribeiro) [reply app/decorators/proxy_rule_decorator.rb]: I liked it. wdyt @Martouta ?

Comment (Martouta) [reply app/decorators/proxy_rule_decorator.rb]: cool :+1:

**Comment by Martouta** [comment]
LGTM :+1: Thanks @guicassolato :1st_place_medal:

**Comment by Martouta** [comment]
Awesome @guicassolato :clap: Unfortunately I cannot approve because I created the PR, but anybody else can :smile:

**Comment by Martouta** [comment]
closing in favor of https://github.com/3scale/porta/pull/1405

**Pull request closed without merging**